docs/src/rfcs/002-shader-stage-interpolants.md

```diff
@@ -0,0 +1,188 @@
+| fragment output | maxFragmentOutputAttachments | 4 | 8
+
+
+### GLSL Linkage mechanism
+
+Historically GLSL used to link shader stage interpolants by name, however in recent years it acquired the `layout(location = N, component = K)` attribute on interpolants. GLSL also allows partial linking where a previous stage can output more data then the stage before.
+
+For pixel shader outputs in modern GLSL, instead of return a value from a function one writes to a global `out` type. And similarly, shader inputs also come from globals.
+
+### Decorators
+
+Similar to DirectX, Vulkan has the following decorations that can be applied to interpolants.
+
+| Vulkan | DirectX equivalent
+| ---- | ---- |
+| - | linear |
+| Flat| nointerpolation
+| NoPerspective| noperspective
+| Centroid | centroid
+| Sample| sample
+
+Vulkan has some additional explanations about these decorations as well:
+
+> A variable decorated with `Flat` will not be interpolated. Instead, it will have the same value for every fragment within a triangle. This value will come from a single provoking vertex. A variable decorated with `Flat` can also be decorated with Centroid or Sample, which will mean the same thing as decorating it only as `Flat`.
+
+## Metal shading language
+
+__Mostly todo.__
+
+One thing to note as a different to SPIR-V, GLSL and HLSL is that MSL doesn't seem to require explicit binding semantics or `Location` specifiers. It's specification has a `[[attribute(N)]]` specifier, but it's only used on tesselation control points and vertex shader inputs. It's specifically not used for communication between shader stages.
+
+# SPIR-V Practicalities
+
+Vertex shaders allow `Location` and `Component` decorations on input variable declarations. The `Location` decoration specifies which vertex input attribute is used to read and interpret the data that a variable will consume. The `Component` decoration allows the location to be more finely specified for scalars and vectors, down to the individual components within a location that are consumed. The components within a location are 0, 1, 2, and 3. A variable starting at component N will consume components N, N+1, N+2, …​ up through its size. For single precision types, it is invalid if the sequence of components gets larger than 3.
+
+
+## Proposal
+
+```rust
+// these `attribute`'s would match the input assembler setup on the CPU
+struct VertexInput {
+    #[attribute(0)]
+    position: Vec3,
+    #[attribute(1)]
+    normal: Vec3,
+    #[attribute(2)]
+    color: Vec3,
+}
+
+// naming this is hard, since vertex shader won't really know what shader
+// stages will run after it until the pipeline state object gets created
+// in vulkan. a tesselation shader may run after this for example,
+// and that decision won't be known by our compiler.
+struct VertexOutput {
+    #[position]
+    position: Vec3,
+    normal: Vec3,
+    color: Vec3,
+}
+
+fn vs_main(#[stage_in] input: &VertexInput) -> VertexOutput {
```

Comment:
Is it valid to not have a `#[spirv(stage_in)]` attribute? No return value?

Reply:
We should probably require both for now, I think there might be some obscure use-cases for vertex shaders so we can try to list them here and see.

- Procedural full-screen quad will require `VertexId`, and so will need to declare that in `#[stage_in]` through something like a `#[vertex_id]` attribute. This would imply that at least `#[attribute(N)]` would be optional.
 - It might be wanted occasionally would want to store out vertex data to a buffer (think debugging tools doing mesh visualization) - in this case you probably wouldn't want to run a pixel shader directly after (or maybe you would). However, this would probably require stream-out support (which we don't have, and probably won't have for a long time since it's quite an obscure feature).
 - Another (much more likely) case would be shadow-map rendering. In this case you don't bind a pixel shader, but still need to output position from the vertex shader.

I think it's safe to require them both, unless somebody has another valid use-case that I didn't think of?